File: movies-api/src/main/java/com/ivanfranchin/moviesapi/movie/model/Movie.java
package com.ivanfranchin.moviesapi.movie.model;

import com.ivanfranchin.moviesapi.movie.dto.CreateMovieRequest;
import com.ivanfranchin.moviesapi.userextra.dto.UpdateMovieRequest;
import lombok.AllArgsConstructor;
import lombok.Data;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

@Data
@Document(collection = "movies")
public class Movie {

    @Id
    private String imdbId;
    private String title;
    private String director;
    private String year;
    private String poster;
    private List<Comment> comments = new ArrayList<>();

    @Data
    @AllArgsConstructor
    public static class Comment {
        private String username;
        private String text;
        private Instant timestamp;
    }

    public static Movie from(CreateMovieRequest createMovieRequest) {
        Movie movie = new Movie();
        movie.setImdbId(createMovieRequest.imdbId());
        movie.setTitle(createMovieRequest.title());
        movie.setDirector(createMovieRequest.director());
        movie.setYear(createMovieRequest.year());
        movie.setPoster(createMovieRequest.poster());
        return movie;
    }

    public static void updateFrom(UpdateMovieRequest updateMovieRequest, Movie movie) {
        if (updateMovieRequest.title() != null) {
            movie.setTitle(updateMovieRequest.title());
        }
        if (updateMovieRequest.director() != null) {
            movie.setDirector(updateMovieRequest.director());
        }
        if (updateMovieRequest.year() != null) {
            movie.setYear(updateMovieRequest.year());
        }
        if (updateMovieRequest.poster() != null) {
            movie.setPoster(updateMovieRequest.poster());
        }
    }
}

File: movies-api/src/main/kotlin/com/ivanfranchin/moviesapi/movie/model/Movie.kt
package com.ivanfranchin.moviesapi.movie.model

import com.ivanfranchin.moviesapi.movie.dto.CreateMovieRequest
import com.ivanfranchin.moviesapi.userextra.dto.MovieDTO
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.Instant

@Document(collection = "movies")
class Movie {

    @Id
    var imdbId: String? = null
    var title: String? = null
    var director: String? = null
    var year: String? = null
    var poster: String? = null
    var comments: MutableList<Comment> = mutableListOf()

    class Comment(
        var username: String,
        var text: String,
        var timestamp: Instant
    )

    companion object {
        fun from(createMovieRequest: CreateMovieRequest): Movie {
            val movie = Movie()
            movie.imdbId = createMovieRequest.imdbId()
            movie.title = createMovieRequest.title()
            movie.director = createMovieRequest.director()
            movie.year = createMovieRequest.year()
            movie.poster = createMovieRequest.poster()
            return movie
        }

        fun updateFrom(updateMovieRequest: MovieDTO.UpdateMovieRequest, movie: Movie) {
            updateMovieRequest.title.let { movie.title = it }
            updateMovieRequest.director.let { movie.director = it }
            updateMovieRequest.year.let { movie.year = it }
            updateMovieRequest.poster.let { movie.poster = it }
        }
    }
}
